### PR TITLE
Feature/filter rest api

### PR DESCRIPTION
**Pull request by alpaka-alpachino**

### CODE REVIEW

**Review by crnopster — DISMISSED** [review]

**Review by maksymbilokur — DISMISSED** [review]

**Review by crnopster — DISMISSED** [review]

File: web/handlers.go [hunk 1]
```diff
@@ -1,3 +1,156 @@
...
+const (
+	inputUserID = "user_id"
+	inputName   = "name"
+)
+
+type repo struct {
+	filters repository.Filters
+}
+
+func newRepo(filters repository.Filters) *repo {
```
Comment (dmitriyminer) [inline web/handlers.go]: This is handler, not repository

File: web/handlers.go [hunk 2]
```diff
@@ -1,3 +1,156 @@
...
+		filters: filters,
+	}
+}
+
+func (repo *repo) AddFilter(w http.ResponseWriter, r *http.Request) {
+	vars := mux.Vars(r)
+	userIDstring := vars[inputUserID]
+
+	userid, err := uuid.Parse(userIDstring)
+	if err != nil {
```
Comment (dmitriyminer) [inline web/handlers.go:39]: Missed errors logging

File: cmd/main.go [hunk 3]
```diff
@@ -25,6 +27,7 @@ func main() {
 
 	configPath := flag.String("cp", "../config", "Path to config file")
 	configName := flag.String("cn", "viper.config", "Name of config file")
+	servicePort := flag.String("p", ":8000", "Service port")
```
Comment (dmitriyminer) [inline cmd/main.go]: Value `:8000` is not a port, it is a format: listen on port 8000 on any interface

File: swagger-api.yaml [hunk 4]
```diff
@@ -100,22 +91,14 @@ paths:
...
         '401':
-          description: >-
-            Failed to validate user credentials. Wrond credentials or user does
-            not exist. Responds with the reason of rejection.
-          content:
-            application/json:
-              schema:
-                $ref: '#/components/schemas/Error'
+        description: >-
+          Failed to validate user credentials. Wrond credentials or user does
```
Comment (dmitriyminer) [inline swagger-api.yaml]: Typo in `Wrond`

File: web/handlers.go [hunk 5]
```diff
@@ -1,3 +1,156 @@
...
+	m := mux.Vars(r)
+	userIDstring := m[inputUserID]
+	userID, err := uuid.Parse(userIDstring)
+	if err != nil {
+		w.WriteHeader(http.StatusUnauthorized)
+		return
+	}
+
+	offsetString := r.URL.Query().Get("offset")
+	offset, _ := strconv.Atoi(offsetString)
```
Comment (dmitriyminer) [inline web/handlers.go]: Missed error handling

File: web/endpoint.go [hunk 6]
```diff
@@ -1,3 +1,24 @@
...
+
+	"github.com/gorilla/mux"
+	"github.com/lvl484/positioning-filter/repository"
+)
+
+func newRouter(filters repository.Filters) *mux.Router {
+	handle := newRepo(filters)
+
+	router := mux.NewRouter()
+	filtersRouter := router.PathPrefix("/v1/router/{user_id}/filters").Subrouter()
```
Comment (dmitriyminer) [inline web/endpoint.go]: Please look into API specification, we have no such route

File: web/handlers.go (hunk 5 above)
Comment (dmitriyminer) [inline web/handlers.go]: Same as https://github.com/lvl484/positioning-filter/pull/34/files#r407080674

File: web/handlers.go [hunk 7]
```diff
@@ -1,3 +1,156 @@
...
+
+	vars := mux.Vars(r)
+
+	if err := json.NewDecoder(r.Body).Decode(&filter); err != nil {
+		w.WriteHeader(http.StatusBadRequest)
+		return
+	}
+	filter.Name = vars[inputName]
+	userIDstring := vars[inputUserID]
+	userUUID, err := uuid.Parse(userIDstring)
```
Comment (dmitriyminer) [inline web/handlers.go:143]: Parsing path parameters should be before decoding request body

File: web/handlers.go [hunk 8]
```diff
@@ -1,3 +1,156 @@
...
+}
+
+func (repo *repo) GetOneFilter(w http.ResponseWriter, r *http.Request) {
+	m := mux.Vars(r)
+	filterName := m[inputName]
+	userIDstring := m[inputUserID]
+	userID, err := uuid.Parse(userIDstring)
+
+	if err != nil {
+		w.WriteHeader(http.StatusUnauthorized)
```
Comment (dmitriyminer) [inline web/handlers.go]: Propose to respond NotFound in this case

File: web/handlers.go [hunk 9]
```diff
@@ -1,3 +1,156 @@
...
+	w.WriteHeader(http.StatusOK)
+}
+
+func (repo *repo) DeleteFilter(w http.ResponseWriter, r *http.Request) {
+	m := mux.Vars(r)
+	filterName := m[inputName]
+	userIDstring := m[inputUserID]
+	userID, err := uuid.Parse(userIDstring)
+	if err != nil {
+		w.WriteHeader(http.StatusUnauthorized)
```
Comment (dmitriyminer) [inline web/handlers.go]: Same as https://github.com/lvl484/positioning-filter/pull/34/files#r407084912

File: swagger-api.yaml (hunk 4 above)
Comment (alpaka-alpachino) [reply swagger-api.yaml]: fixed

File: web/handlers.go (hunk 1 above)
Comment (alpaka-alpachino) [reply web/handlers.go]: fixed)

File: web/handlers.go (hunk 5 above)
Comment (alpaka-alpachino) [reply web/handlers.go]: fixed

Comment (alpaka-alpachino) [reply web/handlers.go]: Same line)) fixed

File: web/handlers.go (hunk 7 above)
Comment (alpaka-alpachino) [reply web/handlers.go:143]: done

File: web/handlers.go (hunk 8 above)
Comment (alpaka-alpachino) [reply web/handlers.go]: yep, thanks

File: web/handlers.go (hunk 9 above)
Comment (alpaka-alpachino) [reply web/handlers.go]: fixed

File: web/endpoint.go (hunk 6 above)
Comment (alpaka-alpachino) [reply web/endpoint.go]: fixed

File: cmd/main.go (hunk 3 above)
Comment (alpaka-alpachino) [reply cmd/main.go]: fixed

File: web/server.go [hunk 10]
```diff
@@ -1,3 +1,34 @@
...
+		&http.Server{
+			Addr:    addr,
+			Handler: router,
+		},
+	}
+}
+
+func (ws *WebServer) Close() error {
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
```
Comment (radikh) [inline web/server.go]: That's enough to use  context.Background() in shutdown, you don't need cancel here.

File: web/handlers_test.go [hunk 11]
```diff
@@ -0,0 +1,230 @@
+// Package web provides managing HTTP protocol.
+// Implement CRUD operations for filters over HTTP API.
+package web
```
Comment (radikh) [inline web/handlers_test.go:3]: Try to reduce duplication in tests. These tests are asserting only expected codes and errors, but not bodies.

File: web/handlers.go (hunk 2 above)
Comment (alpaka-alpachino) [reply web/handlers.go:39]: fixed

File: web/server.go (hunk 10 above)
Comment (alpaka-alpachino) [reply web/server.go]: fixed)

File: swagger-api.yaml [hunk 12]
```diff
@@ -100,22 +91,14 @@ paths:
...
-                $ref: '#/components/schemas/Error'
-        '401':
-          description: >-
-            Failed to validate user credentials. Wrond credentials or user does
-            not exist. Responds with the reason of rejection.
-          content:
-            application/json:
-              schema:
-                $ref: '#/components/schemas/Error'
+        '404':
```
Comment (dmitriyminer) [inline swagger-api.yaml]: Is format valid? Please check

File: web/handlers_test.go [hunk 13]
```diff
@@ -0,0 +1,230 @@
...
+	srv := httptest.NewServer(newRouter(filters))
+	defer srv.Close()
+
+	filter := newTestFilter("Name1", "round")
+	filters.EXPECT().Add(filter).Return(nil)
+
+	b, err := json.Marshal(&filter)
+	assert.NoError(t, err)
+
+	urlString := fmt.Sprintf("%s/v1/router/%s/filters/", srv.URL, filter.UserID)
```
Comment (dmitriyminer) [inline web/handlers_test.go]: Please check your tests

File: web/handlers.go (hunk 2 above)
Comment (dmitriyminer) [reply web/handlers.go:39]: You have application logger, you need to use it
https://github.com/lvl484/positioning-filter/blob/ef97f687b9c4fe729417c132c26429435e2eba40/logger/logger.go#L21

File: swagger-api.yaml (hunk 12 above)
Comment (alpaka-alpachino) [reply swagger-api.yaml]: Oh, my bad, fixed

File: web/handlers_test.go (hunk 13 above)
Comment (alpaka-alpachino) [reply web/handlers_test.go]: Thank you, fixed

File: swagger-api.yaml [hunk 14]
```diff
@@ -257,24 +228,18 @@ paths:
...
-            Failed to validate user credentials. Wrond credentials or user does
-            not exist. Responds with the reason of rejection.
-          content:
-            application/json:
-              schema:
-                $ref: '#/components/schemas/Error'
+                $ref: '#/components/schemas/FilterWithId'
         '404':
-          description: No object found with associated id
+        description: >-
```
Comment (dmitriyminer) [inline swagger-api.yaml]: Invalid format

File: web/handlers.go [hunk 15]
```diff
@@ -1,3 +1,178 @@
...
+}
+
+func (handler *handler) GetOneFilter(w http.ResponseWriter, r *http.Request) {
+	m := mux.Vars(r)
+	filterName := m[inputName]
+	userIDstring := m[inputUserID]
+	userID, err := uuid.Parse(userIDstring)
+
+	if err != nil {
+		handler.log.Error(err)
```
Comment (dmitriyminer) [inline web/handlers.go]: It would be great to add more context about error

File: cmd/main.go [hunk 16]
```diff
@@ -64,8 +67,12 @@ func main() {
 		return
 	}
 
+	filters := repository.NewFiltersRepository(db)
+	srv := web.NewWebServer(filters, *serviceAddr)
```
Comment (dmitriyminer) [inline cmd/main.go]: Is it compiled?

File: swagger-api.yaml (hunk 14 above)
Comment (alpaka-alpachino) [reply swagger-api.yaml]: fixed

File: cmd/main.go (hunk 16 above)
Comment (alpaka-alpachino) [reply cmd/main.go]: forgot ctrl+S here >~<, fixed

**Review by dmitriyminer — DISMISSED** [review]

File: web/handlers_test.go (hunk 11 above)
Comment (alpaka-alpachino) [reply web/handlers_test.go:3]: fixed

**Review by crnopster — APPROVED** [review]

**Review by dmitriyminer — APPROVED** [review]

**Review by maksymbilokur — APPROVED** [review]

**Pull request merged**